Include analysis in the UI

djmitche opened the pull request:
This mostly fixes #8.  There's an endless stream of tweaks to make, but let's call this good for the moment.

djmitche commented:
By the way, thoughts for the future here:
 * It'd be really useful to have tests for this analysis stuff.  Most of it is pretty simple, but I had to futz with the overprovisioned workers stuff a bit to get it right, and testing-by-running is not very efficient.  Currently the code is JSX and uses react hooks, so it's difficult to test without webpack and react infrastructure set up.
 * I expect this kind of analysis will eventually be slow, and doing it synchronously in the browser will become problematic.  It'd be great if we could build a framework to do the analysis in a worker thread and show a spinner until each part of the analysis is complete.
 * We'll want to be able to send links to people to show them this-or-that analysis.  Some ideas to that end:
   * Deploy the UI to Netlify
   * Take a URL to a datastore as a URL argument (along with the other URL components detailing which view to show)
   * Run simulations in tasks, with a nice local command to set them up, and producing the datastore as a URL and a nice link to the analysis tool containing that URL.

I may be getting carried away :)

helfi92 commented on package.json:
```diff
@@ -16,11 +16,13 @@
     "chalk": "^4.1.0",
     "commander": "^6.0.0",
     "eslint": "^7.6.0",
+    "moment": "^2.27.0",
```
`moment` is a heavy package for the UI. Thoughts on using `date-fns` instead? https://date-fns.org/docs/Getting-Started

helfi92 replied:
https://bundlephobia.com/result?p=moment@2.27.0
https://bundlephobia.com/result?p=date-fns@2.15.0

helfi92 commented on src/ui/App.jsx:
```diff
@@ -1,17 +1,31 @@
 import { hot } from 'react-hot-loader';
+import {
+  HashRouter as Router,
```
What was the idea behind using `HashRouter` for the routing rather than `BrowserRouter`?

djmitche replied:
I sort of assumed BrowserRouter wouldn't work, but I didn't actually check.. :)

helfi92 commented on src/ui/analysis.jsx:
```diff
@@ -0,0 +1,125 @@
+ *
+ * This is full of `get` methods, which use React hooks to ensure that long calculations are
+ * re-done only when necessary.  All of these functions should only be called within a
+ * render function.
```
What is a render function in react hooks lingo?

djmitche replied:
It's the same as non-hooks lingo -- either `render()` on a component class or, for a pure functional component, the function itself.

```javascript
const MyCompoent = (props) => {
  // call hooks stuff in here
};
```

helfi92 replied:
From https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both, it looks like class components are not encouraged so render() methods won't be required anymore.

djmitche replied:
That sounds good to me!  Hooks are actually quite nice once you get used to them.  Is "render function" the right terminology, then?

helfi92 replied:
It's basically just the return value of the function component. I'm not sure if I would call it render function but I guess that's better.

helfi92 commented on src/ui/analysis.jsx:
```diff
@@ -0,0 +1,125 @@
+        // workers may still yet take a task; and only look at resolved tasks
+        // as a task cannot be running on a shut-down worker.
+        for (let {startedTime, shutdownTime, resolvedTasks} of finalState.shutdownWorkers.values()) {
+          console.log(resolvedTasks.size);
```
`console.log`

helfi92 commented on src/ui/analysis.jsx:
```diff
@@ -0,0 +1,125 @@
+import DataStoreContext from './datastore';
+
+/**
+ * A singleton class for calculating analysis of a run.
```
Importing this class seems to create a new instance every time. Is it really a singleton?

djmitche replied:
I add a `console.log` before the `export default`, and it only gets called once.  That matches my understanding of how importing / requiring works -- a module's definition is only executed once and the result cached by the require machinery.  It's similar to `export default () => {..}` -- every place you `import` or `require` that module will get the identical function object.

helfi92 replied:
You are right here. My mistake.

helfi92 commented on src/ui/App.jsx:
```diff
@@ -1,17 +1,31 @@
+  HashRouter as Router,
+  Switch,
+  Route,
+  Link
```
`Link` seems to be unused here.

helfi92 commented on src/ui/App.jsx:
```diff
@@ -1,17 +1,31 @@
-import datastore from '../../datastore';
+import Home from './views/Home';
+import Summary from './views/Summary';
+import DataStoreContext from './datastore';
```
`DataStoreContext` is unused in this file.

helfi92 commented on src/ui/views/Summary.jsx:
```diff
@@ -0,0 +1,23 @@
+import { hot } from 'react-hot-loader';
+import analysis from '../analysis';
+
+const statistics = [];
```
Unused.

helfi92 commented on src/ui/views/Summary.jsx:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { Link } from "react-router-dom";
```
Unused.

djmitche commented:
The thrown error was because I modified the simple sim in such a way that it didn't shut down all of its workers in the ramp-down phase.  I had been fiddling around trying to get it to generate overprovisioned workers.  I reverted my changes to the simulation.

djmitche commented on .nvmrc:
```diff
@@ -1 +1 @@
-12
+8
```
I think I know why this keeps sneaking in.. I'll fix.
